pkg/metrics/producers/scheduledcapacity/schedules.go

```diff
@@ -0,0 +1,500 @@
+}
+
+// User input is 0-indexed
+func (s *ScheduleSpec) matchesHour(iter *iterativeSchedule) bool {
+	return s.Hours[iter.Hour] || s.Hours[HOUR_ASTERISK]
+}
+
+// User input is 1-indexed
+func (s *ScheduleSpec) matchesMonth(iter *iterativeSchedule) bool {
+	return s.Month[iter.Month-1] || s.Month[MONTH_ASTERISK]
+}
+
+// User input of Day of Month is 1-indexed, Day of Week is 0-indexed
+func (s *ScheduleSpec) matchesDay(iter *iterativeSchedule) bool {
+
+	return (s.DaysOfWeek[iter.Dow] && s.DaysOfMonth[DOM_ASTERISK]) ||
+		(s.DaysOfMonth[iter.Dom-1] && s.DaysOfWeek[DOW_ASTERISK]) ||
+		(s.DaysOfWeek[iter.Dow] || s.DaysOfMonth[iter.Dom-1]) ||
+		(s.DaysOfWeek[DOW_ASTERISK] && s.DaysOfMonth[DOM_ASTERISK])
+}
+
+type iterativeSchedule struct {
+	Year   int
+	Minute int
+	Hour   int
+	Dom    int
+	// TODO Change to some struct that takes in string as well
+	Month int
+	// TODO Change to some struct that takes in string as well
+	Dow int
+}
+
+func (i *iterativeSchedule) convertToVolatileTime() *apis.VolatileTime {
+	return &apis.VolatileTime{Inner: metav1.Time{Time: time.Date(i.Year, time.Month(i.Month), i.Hour, i.Dom, i.Minute, 0, 0, time.Local)}}
+}
+
+// returns true if the time i represents comes before t
+func (i *iterativeSchedule) isBefore(t time.Time) bool {
+	if i.Year == t.Year() {
+		if i.Month == int(t.Month()) {
+			if i.Dom == t.Day() {
+				if i.Hour == t.Hour() {
+					if i.Minute == t.Minute() {
+						return true
+					}
+					return i.Minute < t.Minute()
+				}
+				return i.Hour < t.Hour()
+			}
+			return i.Dom < t.Day()
+		}
+		return i.Month < int(t.Month())
+	}
+	return i.Year < t.Year()
+}
+
+func (i *iterativeSchedule) isAfter(t time.Time) bool {
+	if i.Year == t.Year() {
+		if i.Month == int(t.Month()) {
+			if i.Dom == t.Day() {
+				if i.Hour == t.Hour() {
+					if i.Minute == t.Minute() {
+						return true
+					}
+					return i.Minute > t.Minute()
+				}
+				return i.Hour > t.Hour()
+			}
+			return i.Dom > t.Day()
+		}
+		return i.Month > int(t.Month())
+	}
+	return i.Year > t.Year()
+}
+func (i *iterativeSchedule) getWeekday() int {
+	return int(time.Date(i.Year, time.Month(i.Month), i.Dom, 0, 0, 0, 0, time.Local).Weekday())
+}
+
+// returns month and int so that reduceDay can get the proper amount of days
+// sign is true if iterating forward in time, negative if iterating backwards in time
+func (i *iterativeSchedule) incrementMonth(sign bool) {
+	if sign {
+		if i.Month == 12 {
+			i.Month = 1
+			i.Year += 1
+		} else {
+			i.Month += 1
+		}
+		i.Dom = 1
+		i.Hour = 0
+		i.Minute = 0
+		i.Dow = i.getWeekday()
+	} else {
+		if i.Month == 1 {
+			i.Month = 12
+			i.Year -= 1
+		} else {
+			i.Month -= 1
+		}
+		i.Dom = getDaysInMonth(i.Month, i.Year)
+		i.Hour = 23
+		i.Minute = 59
+		i.Dow = i.getWeekday()
+	}
+}
+
+// sign is true if iterating forward in time, negative if iterating backwards in time
+func (i *iterativeSchedule) incrementDay(sign bool) {
```

Comment:
Just a general comment ... seems like you really shouldn't have to do this kind of date math stuff (which is also very tricky to get right) seems like should be some way to get an existing battle tested Time/Date (standard?) library to do the heavy lifting for you.